lib/topics_filter.rb

```diff
@@ -368,6 +370,26 @@ def tag_ids_from_tag_names(tag_names)
     all_tag_ids
   end
 
+  def filter_tag_groups(values:)
+    values.each do |key_prefix, tag_groups|
+      should_exclude = "NOT" if key_prefix == "-"
+      tag_group_ids = TagGroup.where(name: tag_groups).pluck(:id)
+      @scope =
+        @scope
+          .joins(:tags)
+          .where(
+            "tags.id #{should_exclude} IN (SELECT tag_id FROM tag_group_memberships WHERE tag_group_id IN (?))",
+            tag_group_ids,
+          )
```

Comment:
I may have missed something here, but I'm not sure.

I would love some feedback on getting this SQL query right.

If all topics have tags, they behave correctly, but if a topic does not have any tag, it doesn't bring it.

My second test case, is about this scenario

Reply:
> If all topics have tags, they behave correctly, but if a topic does not have any tag, it doesn't bring it.

I think you need to use a `left join` for the use case where topics have no tags. A `joins(:tags)` clause will always return topics that have associated records in the `tags` table. 

Might be easier to write the `NOT` clause separately, because that is the use case where you would want topics without given tags.